File: 230_kth_smallest_element_in_BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        """
            Time Complexity - O(k)
            Space Complexity - O(h)
            'h' is the height of the tree
        """
        # edge case
        if not root:
            return -1
        # inorder traversal
        # LDR
        stack = []
        while root or stack:
            if root:
                stack.append(root)
                root = root.left
            else:
                cur = stack.pop()
                k -= 1
                if k == 0:
                    return root.val
                root = cur.right

    def kthSmallest(self, root: TreeNode, k: int) -> int:
        if not root:
            return -1
        self.result = None
        self.k = k
        self._inorder_traveral(root)
        return self.result

    def _inorder_traveral(self, root):
        # base
        if not root:
            return
        # logic
        self._inorder_traveral(root.left)
        self.k -= 1
        if self.k == 0:
            self.result = root.val
            return
        self._inorder_traveral(root.right)

File: 230_kthSmallestElementInBST.cpp
// Time Complexity : O(H+K) where H is the height of the tree and K is the Kth smallest element
// Space Complexity :O(H)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

class Solution {
public:
    int cnt=0;
    int result=0;
    int kthSmallest(TreeNode* root, int k) {
        if(root == NULL)
            return -1;
        result = -1;
        inorder(root, k);
        return result;
    }
    
    void inorder(TreeNode* root, int k){
        //base
        if(root == NULL)
            return;
            
        //logic
        inorder(root->left, k);
        if(++cnt == k)
            result = root->val;
        inorder(root->right, k);
    }
    
};


class Solution {
public:
    int cnt=0;
    int result=0;
    int kthSmallest(TreeNode* root, int k) {
        if(root == NULL)
            return -1;
        stack<TreeNode*> s;
        while(root != NULL || !s.empty()){
            while(root != NULL){
                s.push(root);
                root = root->left;
            }
            root = s.top();
            s.pop();
            k--;
            if(k == 0)
                return root->val;
            root = root->right;
        }
        return -1;
    }
    
};
